### PR TITLE
Implemented Go routines for health checks

### PR DESCRIPTION
**Pull request by janithht**
This pull request implements parallel go routines to run health checks for each upstream.
#Issues:
- Handling Different Numbers of Servers in Each Upstream
- Respecting Unique Health Check Intervals

### CODE REVIEW

File: internal/healthchecks/healthCheck.go [hunk 1]
```diff
@@ -17,19 +22,35 @@ func checkServerHealth(server string, healthCheckConfig config.HealthCheck) bool
...
 }
 
-func PerformHealthChecks(config *config.Config) {
+func worker(id int, tasks <-chan HealthCheckTask) {
+	for task := range tasks {
+		if !checkServerHealth(task.Server, task.HealthCheckConfig) {
+			log.Printf("[Worker %d] Server %s failed health check, removing from pool\n", id, task.Server)
+		} else {
+			log.Printf("[Worker %d] Server %s passed health check\n", id, task.Server)
+		}
```
Comment (binura-g) [inline internal/healthchecks/healthCheck.go]: it's not enough to log this information right?

You need to update the health status somewhere

File: internal/healthchecks/healthCheck.go [hunk 2]
```diff
@@ -17,19 +22,35 @@ func checkServerHealth(server string, healthCheckConfig config.HealthCheck) bool
...
+	}
+}
+
+func PerformHealthChecks(cfg *config.Config) {
```
Comment (binura-g) [inline internal/healthchecks/healthCheck.go]: we need to start passing in a context.Context in go. Read up on it.

Comment (binura-g) [reply internal/healthchecks/healthCheck.go]: https://www.youtube.com/watch?v=LSzR0VEraWw

File: internal/healthchecks/healthCheck.go [hunk 3]
```diff
@@ -17,19 +22,35 @@ func checkServerHealth(server string, healthCheckConfig config.HealthCheck) bool
...
+	const numWorkers = 10
+	tasks := make(chan HealthCheckTask, 100)
+
+	for i := 0; i < numWorkers; i++ {
+		go worker(i, tasks)
+	}
+
 	for {
 		fmt.Println()
```
Comment (binura-g) [inline internal/healthchecks/healthCheck.go]: why is there a println here

File: internal/healthchecks/healthCheck.go [hunk 4]
```diff
@@ -17,19 +22,35 @@ func checkServerHealth(server string, healthCheckConfig config.HealthCheck) bool
...
-					upstream.Servers = append(upstream.Servers[:i], upstream.Servers[i+1:]...)
+		for _, upstream := range cfg.Upstreams {
+			for _, server := range upstream.Servers {
+				task := HealthCheckTask{
+					Server:            server,
+					HealthCheckConfig: upstream.HealthCheck,
 				}
+				tasks <- task
 			}
+			time.Sleep(upstream.HealthCheck.Interval)
```
Comment (binura-g) [inline internal/healthchecks/healthCheck.go:69]: really think about this... is this doing what you think it should be doing?

Comment (binura-g) [reply internal/healthchecks/healthCheck.go:69]: try to think about how to write tests for this part so you can validate the following:

- that is checks all the upstreams - ensuring the interval provided in the config is honoured.
- and that one upstreams healthcheck interval does not affect the others.

You can to some extend validate these by adding more upstreams in the config, printing the timestamps and checking the time differences manually. But you need to figure out to how actually write this as a test suite.

File: internal/healthchecks/healthCheck.go (hunk 3 above)
Comment (janithht) [reply internal/healthchecks/healthCheck.go]: Just to separate the healthcheck lines

File: internal/healthchecks/healthCheck.go (hunk 1 above)
Comment (janithht) [reply internal/healthchecks/healthCheck.go]: Used a map to store the health stasus of each server

**Pull request merged**